### PR TITLE
Dataset.plot.quiver

### LINKED ISSUE #4373: Add Dataset.plot.quiver
I think it would be nice to add a quiver plot function. I got this far in my current project:

``` python
@xarray.plot.dataset_plot._dsplot
def quiver(ds, x, y, ax, u, v, **kwargs):
    from xarray import broadcast

    if x is None or y is None or u is None or v is None:
        raise ValueError("Must specify x, y, u, v for quiver plots.")

    # matplotlib autoscaling algorithm
    scale = kwargs.pop("scale", None)
    if scale is None:
        npts = ds.dims[x] * ds.dims[y]
        # crude auto-scaling
        # scale is typical arrow length as a multiple of the arrow width
        scale = (
            1.8 * ds.to_array().median().values * np.maximum(10, np.sqrt(npts))
        )  # / span

    ds = ds.squeeze()
    x, y, u, v = broadcast(ds[x], ds[y], ds[u], ds[v])

    # no support for these kwargs right now.
    kwargs.pop("cmap_params")
    kwargs.pop("hue")
    kwargs.pop("hue_style")
    hdl = ax.quiver(x.values, y.values, u.values, v.values, scale=scale, **kwargs)

    return hdl
```

The autoscaling logic is quite crude; I tried to copy what matplotlib does but got somewhat confused. To get faceting to work properly, we'll need to estimate `scale` around here: https://github.com/pydata/xarray/blob/a36d0a1d4657c848dcdd76d0ecb9c783ad464057/xarray/plot/facetgrid.py#L339-L343

### PR DESCRIPTION
**Pull request by dcherian**
- [x] Closes #4373 
 - [x] Tests added
 - [x] Passes `isort . && black . && mypy . && flake8`
 - [x] User visible changes (including notable bug fixes) are documented in `whats-new.rst`
 - [x] New functions/methods are listed in `api.rst`

I could use some help with  adding tests and parameter checking if someone wants to help :)

``` python
import numpy as np
import xarray as xr

ds = xr.Dataset()
ds.coords["x"] = ("x", np.arange(10))
ds.coords["y"] = ("y", np.arange(20))
ds.coords["t"] = ("t", np.arange(4))
ds["u"] = np.sin((ds.x - 5) / 5) * np.sin((ds.y - 10) / 10)
ds["v"] = np.sin((ds.x - 5) / 5) * np.cos((ds.y - 10) / 10)

ds = ds * 2*np.cos((ds.t) * 2 * 3.14 /0.75)

ds["u"].attrs["units"] = "m/s"
ds["mag"] = np.hypot(ds.u, ds.v)

ds.mag.plot(col="t", x="x")

fg = ds.plot.quiver(x="x", y="y", u="u", v="v", col="t", hue="mag")
```

image
image

### CODE REVIEW

**Comment by dcherian** [comment]
This is ready for a review. I don't I can fix the quiverkey without https://github.com/matplotlib/matplotlib/issues/18530 . This is still useful for a quick look at the data.

File: xarray/plot/dataset_plot.py [hunk 1]
```diff
@@ -48,11 +49,24 @@ def _infer_meta_data(ds, x, y, hue, hue_style, add_guide):
...
         add_legend = False
         add_colorbar = False
 
+    if (add_guide or add_guide is None) and funcname == "quiver":
+        add_quiverkey = True
+        if hue:
+            add_colorbar = True
+            if not hue_style:
+                hue_style = "continuous"
+            elif hue_style != "continuous":
```
Comment (mathause) [inline xarray/plot/dataset_plot.py:63]: Could you test for this?

File: xarray/plot/dataset_plot.py [hunk 2]
```diff
@@ -344,13 +367,34 @@ def newplotfunc(
                 cbar_kwargs["label"] = meta_data.get("hue_label", None)
             _add_colorbar(primitive, ax, cbar_ax, cbar_kwargs, cmap_params)
 
+        if meta_data["add_quiverkey"]:
+            magnitude = _get_nice_quiver_magnitude(ds[u], ds[v])
+            units = ds[u].attrs.get("units", "")
```
Comment (mathause) [inline xarray/plot/dataset_plot.py:372]: could you test the unit string of u is used?

**Review by mathause — COMMENTED** [review]
Looks good. The plotting code is not getting simpler 🙃  but it's so useful!

* Why is scale required for facetgrid and not for a single panel?
* Should it emit a Warning/ Error if `u` or `v` are given and `funcname != "quiver"`

File: xarray/plot/dataset_plot.py (hunk 1 above)
Comment (dcherian) [reply xarray/plot/dataset_plot.py:63]: done.

File: xarray/plot/dataset_plot.py (hunk 2 above)
Comment (dcherian) [reply xarray/plot/dataset_plot.py:372]: done.

**Comment by dcherian** [comment]
> Why is scale required for facetgrid and not for a single panel?

matplotlib will make a nice choice for the single panel. With facetgrid, the scale will be different for each panel if we don't specify it. And i had trouble porting the matplotlib algorithm. We can always add this in the future.

> Should it emit a Warning/ Error if u or v are given and funcname != "quiver"

Good idea. Done.

File: xarray/plot/facetgrid.py [hunk 3]
```diff
@@ -195,7 +196,11 @@ def __init__(
         self.axes = axes
         self.row_names = row_names
         self.col_names = col_names
+
+        # guides
         self.figlegend = None
+        self.quiverkey = None
+        self.cbar = None
```
Comment (dcherian) [inline xarray/plot/facetgrid.py:203]: technically backwards incompatible but I'm not sure it affects anyone.

File: doc/plotting.rst [hunk 4]
```diff
@@ -762,6 +765,26 @@ Faceting is also possible
 
 For more advanced scatter plots, we recommend converting the relevant data variables to a pandas DataFrame and using the extensive plotting capabilities of ``seaborn``.
 
+Quiver
+~~~~~~
+
+Visualizing vector fields using a quiver plot is easy.
```
Comment (mathause) [inline doc/plotting.rst:771]:
```suggestion
Visualizing vector fields is supported with quiver plots:
```

File: doc/plotting.rst [hunk 5]
```diff
@@ -762,6 +765,26 @@ Faceting is also possible
...
+
+.. ipython:: python
+    :okwarning:
+
+    @savefig ds_simple_quiver.png
+    ds.isel(w=1, z=1).plot.quiver(x="x", y="y", u="A", v="B")
+
+
+Again, faceting is also possible
```
Comment (mathause) [inline doc/plotting.rst:780]:
```suggestion
where ``u`` and ``v`` denote the x and y direction components of the arrow vectors. Again, faceting is also possible:
```

File: doc/plotting.rst [hunk 6]
```diff
@@ -762,6 +765,26 @@ Faceting is also possible
...
+
+.. ipython:: python
+    :okwarning:
+
+    @savefig ds_facet_quiver.png
+    ds.plot.quiver(x="x", y="y", u="A", v="B", col="w", row="z", scale=4)
+
```
Comment (mathause) [inline doc/plotting.rst:787]:
```suggestion

The ``scale`` is required for facet plots. The scale determines the number of data units per arrow length unit, i.e. a smaller scale parameter makes the arrow longer.
```

**Review by mathause — APPROVED** [review]
Looks good (but I don't have experience with `plt.quiver`).

**Comment by dcherian** [comment]
I'll merge on Friday if there are no more comments.

**Pull request merged**